### ISSUE TITLE
Race condition during process handoff caused by handoff occuring prior to registry update

### ISSUE THREAD

**Issue by dkulchenko** [derekkraan/horde#188]
(on the master branch)

I'm seeing an issue that occurs when I kill one of the nodes in a two-node cluster, where on the still alive node, I'm seeing:

```
[error] pid=<0.1497.0> file=gen_server.erl function=error_info/7 line=889  GenServer Kiwi.DistributedSupervisor terminating
** (MatchError) no match of right hand side value: {:ignore, %Horde.DynamicSupervisorImpl{distribution_strategy: Horde.UniformDistribution, local_process_count: 13, members: %{{Kiwi.DistributedSupervisor, :"a@127.0.0.1"} => 1, {Kiwi.DistributedSupervisor, :"b@127.0.0.1"} => 1}, members_info: %{{Kiwi.DistributedSupervisor, :"a@127.0.0.1"} => %Horde.DynamicSupervisor.Member{name: {Kiwi.DistributedSupervisor, :"a@127.0.0.1"}, status: :alive}, {Kiwi.DistributedSupervisor, :"b@127.0.0.1"} => %Horde.DynamicSupervisor.Member{name: {Kiwi.DistributedSupervisor, :"b@127.0.0.1"}, status: :dead}}, name: Kiwi.DistributedSupervisor, name_to_supervisor_ref: %{{Kiwi.DistributedSupervisor, :"a@127.0.0.1"} => #Reference<0.765265296.2688548871.70159>}, process_pid_to_id: %{nil => 138912447372216481172297720070098[....]strategy: :one_for_one, distribution_strategy: Horde.UniformDistribution, process_redistribution: :passive, members: :auto], supervisor_ref_to_name: %{#Reference<0.765265296.2688548871.70159> => {Kiwi.DistributedSupervisor, :"a@127.0.0.1"}}, waiting_for_quorum: []}}
    (horde) lib/horde/dynamic_supervisor_impl.ex:414: anonymous fn/3 in Horde.DynamicSupervisorImpl.handoff_processes/1
    (elixir) lib/enum.ex:1948: Enum."-reduce/3-lists^foldl/2-0-"/3
    (horde) lib/horde/dynamic_supervisor_impl.ex:378: Horde.DynamicSupervisorImpl.handle_info/2
    (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:711: :gen_server.handle_msg/6
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message: {:crdt_update, [{:add, {:member_node_info, {Kiwi.DistributedSupervisor, :"b@127.0.0.1"}}, %Horde.DynamicSupervisor.Member{name: {Kiwi.DistributedSupervisor, :"b@127.0.0.1"}, status: :dead}}]}
```

I've tracked down the issue to the fact that the process handoff in response to the node death is happening just prior to the corresponding registry update, causing the registry to deny the new process's registration via tuple, saying that the process is already started (on the dead node B).

In other words, the process is running on node B, node B dies, node A process supervisor gets the CRDT update and attempts to restart the process on node A, but the registry still thinks it's running on node B because the registry CRDT update hasn't been processed just yet (but will be in literally microseconds).

**Comment by dkulchenko** [comment]
Kind of a silly workaround, but changing the sync_interval on the Registry to 100ms and keeping the Supervisor at 300ms makes the problem go away, although this still feels like a bug of some sort.

**Comment by dkulchenko** [comment]
Sadly the workaround didn't work in production, so working around it with https://github.com/dkulchenko/horde/commit/1f907e4e4b191bcde8efa251b13874a52888ed5f instead for now.

**Comment by derekkraan** [comment]
@dkulchenko fundamentally there seems to be a race condition here between the Supervisor and the Registry. Locally, this race condition does not occur since both the registry and the supervisor get notified of a process dying simultaneously, so the registry gets the exit signal in its process mailbox before any message to register the replacement process can reach it.

In a multi-node setup, we can't guarantee this, so sometimes the Registry (on another node) hasn't heard that the process has died before the Supervisor decides to restart it. This is especially noticeable in Horde since we only sync once in so many milliseconds. You can try to set the sync_interval lower, but at the end of the day, if the update from the supervisor comes in first, then the race condition will be triggered.

So we need another approach.

A) One option would be to serialize all updates through a single delta CRDT. I'm not sure if this is really a good idea though.

B) Another option is ... to immediately (re-)broadcast exit signals to all other registry nodes in the cluster.

C) Yet another option: always allow updates to clobber a remote registration. In the case that the remote process does exist, this would trigger the conflict resolution mechanism and kill that remote process. In the case that the remote process does not exist anymore, then it will have been the correct choice.

I think at the end of the day, your approach of allowing `:ignore` needs to happen anyways, since this _is_ a legitimate return value of `start_child`, but it's not a real "solution" to this problem, since it leads to lost children (reg B says "nope, already registered", so sup B drops the process from the CRDT, then reg B gets the message to deregister the process and it's completely gone from the system).

I am sort of leaning towards solution C at the present time, but I think I'll have to think about this a little longer before I am sure.

Side note: I think this race condition is what is causing other complaints of processes disappearing from Horde.

**Comment by derekkraan** [comment]
On second thought, option C won't solve anything, since nothing will stop the supervisor from restarting a process that has been killed by the registry's conflict resolution mechanism.

**Comment by derekkraan** [comment]
Perhaps C will help solve the problem. It should work if we set our processes to be transient, and then intercept the exit signal for losing a conflict and make the process stop with `{:exit, :normal}`. I _think_ this should prevent registration loops and allow us to also handle this race condition.

**Comment by derekkraan** [comment]
Hang on a second, how can this be happening? In `Horde.Registry`, we check whether a pid is still alive (even remote ones) before allowing one to be registered. Ignoring the fact for a second that this will probably lead to degenerate performance in a netsplit case (which is not what we want), it should prevent this type of race condition from occurring.

**Comment by dkulchenko** [comment]
I like option C as well. As for how it's happening, I'm not entirely sure either, but I do have a minimal test case for you here: https://github.com/dkulchenko/horderace that triggers the issue each time.

Repro steps:

1. Open two terminals and cd into the horderace git clone
2. In one terminal run, `iex --name a@127.0.0.1 -S mix`
3. Wait a sec for the app to fully start
4. In the other, run, `iex --name b@127.0.0.1 -S mix`
5. Now in the *first* terminal, the one running a@, press Ctrl-C and then `a` to abort the process
6. You'll see this issue's stacktrace in the second terminal

**Comment by derekkraan** [comment]
Thanks for your test case @dkulchenko! I was able to translate it into a test, and I am pretty confident I've also fixed the bug now. I have merged it to `master`, could you please confirm that this has solved the problem?

**Comment by dkulchenko** [comment]
Passes the test case now, thank you :)

I'm a bit concerned about 5e45f1f78c54ed8ba424ab2f64a694bd625f1b14 though - is there a way that matching against any return can result in a situation where the process is lost in the cluster? For example, add_child returns :ignore, and so the process doesn't get started, but then the original node doesn't have the process anymore either because it's shutting down, so the process disappears entirely, like you mentioned in your first comment above.

Also, what if add_child returns `{{:error, _}, state}`? Could that result in process loss during handoff?

**Comment by derekkraan** [comment]
Yes, if a child process fails to start, then it will be discarded ("lost"). I believe this is consistent with the behaviour of `DynamicSupervisor` (https://hexdocs.pm/elixir/DynamicSupervisor.html#start_child/2)

**Comment by dkulchenko** [comment]
Hmm - but is it possible to silently lose a process during handoff because of the race condition from the original issue? e.g. start_child reports :ignore so it doesn't get started on the new node, but then it also disappears from the old node, so it disappears from the cluster entirely.

The fix will make it so the supervisor doesn't crash in response to an :ignore during handoff, but I'm trying to figure out what will prevent process loss in this situation.

**Comment by derekkraan** [comment]
I'm not convinced there is a race condition yet. I am fairly certain that this is the correct behaviour though. If there is a race condition then we should see what we can do to fix that.

It doesn't "silently" drop processes anyways, since the process (ie your code) is telling it to `:ignore` the process.
